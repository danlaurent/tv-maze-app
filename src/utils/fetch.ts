export const fetchService = async <T>(url: string): Promise<T> => {
  try {
    const data = await fetch(url);
    const response = await data.json();
    return response;
  } catch (error) {
    throw error;
  }
};
