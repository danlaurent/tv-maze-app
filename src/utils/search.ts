import { TSearchedShowList } from '../interface/shows';

export const handleSearchResults = (results: TSearchedShowList) =>
  results.map((result) => result.show);
