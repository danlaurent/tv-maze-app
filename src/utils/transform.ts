import { IEpisode, ISchedule } from '../interface/shows';

export const cleanSummary = (summary: string) =>
  summary.replace(/(<([^>]+)>)/gi, '');

export const humanizeMetaInfo = (metainfo: string[]) => metainfo.join(' | ');

export const handleShowSeasons = (episodes: IEpisode[]) => {
  const seasonNumbers = episodes.map((episode) => episode.season);
  const seasons = [...new Set(seasonNumbers)];

  const showSeasons = seasons.map((item) => ({
    season: item,
    episodes: episodes.filter((episode) => episode.season === item),
  }));

  return showSeasons;
};

export const handleSchedule = (schedule: ISchedule) => {
  if (schedule?.time && schedule?.days) {
    return `${schedule.time} on ${humanizeMetaInfo(schedule.days)}`;
  }
  return undefined;
};
