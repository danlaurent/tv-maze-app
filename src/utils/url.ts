export const getUrlWithOptions = (url: string, options?: any) => {
  if (options) {
    const params = Object.keys(options)
      .map((key) => `${key}=${options[key]}`)
      .join('&');
    return `${url}?${params}`;
  }
  return url;
};
