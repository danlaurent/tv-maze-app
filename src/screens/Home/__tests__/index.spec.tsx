import React from 'react';
import {
  render,
  RenderAPI,
  fireEvent,
  act,
} from '@testing-library/react-native';
import HomeScreen from '..';
import * as fetchUtils from '../../../utils/fetch';

const navigationMock: any = {
  navigate: jest.fn(),
  goBack: jest.fn(),
};

const routeMock: any = {
  params: {
    showId: 1,
  },
};

const mockShowResponse = [
  {
    id: 1,
    name: 'testSuccess',
    genres: ['Drama', 'Science-Fiction', 'Thriller'],
    rating: {
      average: 10,
    },
  },
  {
    id: 2,
    name: 'testSuccess2',
    genres: ['Drama', 'Science-Fiction', 'Thriller'],
    rating: {
      average: 10,
    },
  },
];

jest.mock('../../../hooks/useFetch', () => ({
  useFetch: () => ({
    data: mockShowResponse,
    loading: true,
  }),
}));

const mockFetchService = () => {
  jest
    .spyOn(fetchUtils, 'fetchService')
    .mockResolvedValue([{ score: 1, show: { ...mockShowResponse[0] } }]);
};

describe('HomeScreen', () => {
  let tree: RenderAPI;
  beforeEach(() => {
    tree = render(<HomeScreen navigation={navigationMock} route={routeMock} />);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const snapshot = tree.toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('renders the loading', () => {
    expect(tree.getByTestId('HomeScreenLoading')).toBeDefined();
  });

  describe('When a show is pressed', () => {
    beforeEach(async () => {
      await fireEvent.press(tree.getByTestId('HomeScreenShowList_1'));
    });

    it('navigates to the ShowDetails screen', () => {
      expect(navigationMock.navigate).toHaveBeenCalledWith(
        'ShowDetailsScreen',
        { showId: 1 }
      );
    });
  });

  describe('When the user types on search input', () => {
    beforeEach(async () => {
      await act(async () => {
        await fireEvent.changeText(
          tree.getByTestId('HomeScreenSearchBar_searchInput'),
          'Test'
        );
      });
    });
    it('changes the input text', () => {
      const searchInputProps = tree.getByTestId(
        'HomeScreenSearchBar_searchInput_textInput'
      ).props;
      expect(searchInputProps.value).toBe('Test');
    });
  });

  describe('When search is pressed', () => {
    beforeEach(async () => {
      mockFetchService();
      await act(async () => {
        await fireEvent.press(tree.getByTestId('HomeScreenSearchBar_button'));
      });
    });
    it('searches for the show', () => {
      expect(fetchUtils.fetchService).toBeCalled();
    });

    it('shows the search results list', () => {
      expect(tree.getByTestId('HomeScreenSearchList')).toBeDefined();
    });

    describe('When a searched show is pressed', () => {
      beforeEach(async () => {
        await fireEvent.press(tree.getByTestId('HomeScreenSearchList_1'));
      });

      it('navigates to the ShowDetails screen', () => {
        expect(navigationMock.navigate).toHaveBeenCalledWith(
          'ShowDetailsScreen',
          { showId: 1 }
        );
      });
    });

    describe('When the cancel button is pressed', () => {
      it('clears the search field', async () => {
        await act(async () => {
          await fireEvent.press(tree.getByTestId('HomeScreenSearchBar_button'));
        });
        const searchInputProps = tree.getByTestId(
          'HomeScreenSearchBar_searchInput_textInput'
        ).props;
        expect(searchInputProps.value).toBe('');
      });
    });
  });
});
