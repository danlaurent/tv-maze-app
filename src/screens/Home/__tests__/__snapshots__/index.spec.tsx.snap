// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#191a2f",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#191a2f",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginTop": null,
          "paddingHorizontal": 20,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
        testID="HomeScreenSearchBar"
      >
        <View
          style={
            Object {
              "flex": 1,
              "marginRight": 12,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#121322",
                "borderColor": "#0c0d18",
                "borderRadius": 20,
                "borderWidth": 1,
                "flexDirection": "row",
                "height": 60,
                "padding": 12,
              }
            }
            testID="HomeScreenSearchBar_searchInput"
          >
            <Text />
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              placeholder="Show name"
              placeholderTextColor="#6d6f7e"
              rejectResponderTermination={true}
              style={
                Object {
                  "color": "white",
                  "marginLeft": 12,
                }
              }
              testID="HomeScreenSearchBar_searchInput_textInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 14,
              "height": 48,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
          testID="HomeScreenSearchBar_button"
        >
          <Text
            style={
              Object {
                "color": "#c6c7d2",
                "fontSize": 14,
                "textAlign": "center",
              }
            }
          >
            Search
          </Text>
        </View>
      </View>
    </View>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "padding": 20,
        }
      }
      data={
        Array [
          Object {
            "genres": Array [
              "Drama",
              "Science-Fiction",
              "Thriller",
            ],
            "id": 1,
            "name": "testSuccess",
            "rating": Object {
              "average": 10,
            },
          },
          Object {
            "genres": Array [
              "Drama",
              "Science-Fiction",
              "Thriller",
            ],
            "id": 2,
            "name": "testSuccess2",
            "rating": Object {
              "average": 10,
            },
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={0.9}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      testID="HomeScreenShowList"
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "paddingVertical": 8,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flexDirection": "row",
                  "opacity": 1,
                }
              }
              testID="HomeScreenShowList_1"
            >
              <View
                style={
                  Object {
                    "borderRadius": 4,
                    "height": 147,
                    "marginRight": 12,
                    "width": 105,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "stretch",
                      "backgroundColor": "#131425",
                      "borderRadius": 4,
                      "flex": 1,
                      "justifyContent": "center",
                    }
                  }
                  testID="undefined_imagePlaceholder"
                >
                  <Text />
                </View>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 24,
                      "fontWeight": "700",
                      "marginBottom": 8,
                      "marginTop": 16,
                    }
                  }
                >
                  testSuccess
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#6d6f7e",
                      "marginBottom": 8,
                      "textAlign": "center",
                    }
                  }
                  testID="HomeScreenShowList_1_detailsMainInfo_subtitle"
                >
                  Drama | Science-Fiction | Thriller
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Object {
                        "color": "#c6c7d2",
                        "fontSize": 16,
                        "fontWeight": "700",
                        "marginLeft": 8,
                      }
                    }
                    testID="HomeScreenShowList_1_detailsMainInfo_rating"
                  >
                    10
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "paddingVertical": 8,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flexDirection": "row",
                  "opacity": 1,
                }
              }
              testID="HomeScreenShowList_2"
            >
              <View
                style={
                  Object {
                    "borderRadius": 4,
                    "height": 147,
                    "marginRight": 12,
                    "width": 105,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "stretch",
                      "backgroundColor": "#131425",
                      "borderRadius": 4,
                      "flex": 1,
                      "justifyContent": "center",
                    }
                  }
                  testID="undefined_imagePlaceholder"
                >
                  <Text />
                </View>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 24,
                      "fontWeight": "700",
                      "marginBottom": 8,
                      "marginTop": 16,
                    }
                  }
                >
                  testSuccess2
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#6d6f7e",
                      "marginBottom": 8,
                      "textAlign": "center",
                    }
                  }
                  testID="HomeScreenShowList_2_detailsMainInfo_subtitle"
                >
                  Drama | Science-Fiction | Thriller
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Object {
                        "color": "#c6c7d2",
                        "fontSize": 16,
                        "fontWeight": "700",
                        "marginLeft": 8,
                      }
                    }
                    testID="HomeScreenShowList_2_detailsMainInfo_rating"
                  >
                    10
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginBottom": 32,
        }
      }
      testID="HomeScreenLoading"
    >
      <ActivityIndicator
        animating={true}
        color="white"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
